### PR TITLE
[wip] adjust target model to support 5x4

### PR DESCRIPTION
**Pull request by makslevental**
This PR parameterizes `AIETargetModel` with a `bool virtualized`, which is then queried when checking `isShimNOCTile(col, row) || isShimPLTile(col, row)`. Note, I have only updated `IPUTargetModel` to the correct result because that is the only one I am familiar with (the rest fallback to `virtualized=true`, i.e., the current behavior).

Note, this PR touched a lot of lines only because the models were `static` prior and had to be updated to `shared_ptr`s.

fixes https://github.com/Xilinx/mlir-aie/issues/1094

### CODE REVIEW

**Comment by jgmelber** [comment]
Do you have a new test for 5 columns?

**Comment by makslevental** [comment]
> Do you have a new test for 5 columns?

In another branch yes

File: lib/Dialect/AIE/Transforms/AIECanonicalizeDevice.cpp [hunk 1]
```diff
@@ -39,8 +39,8 @@ struct AIECanonicalizeDevicePass
 
     Location location = builder.getUnknownLoc();
     auto deviceOp = builder.create<DeviceOp>(
-        location,
-        AIEDeviceAttr::get(builder.getContext(), AIEDevice::xcvc1902));
+        location, AIEDeviceAttr::get(builder.getContext(), AIEDevice::xcvc1902),
+        /*virtualized*/ builder.getBoolAttr(true));
```
Comment (stephenneuendorffer) [inline lib/Dialect/AIE/Transforms/AIECanonicalizeDevice.cpp:43]: This seems wrong, because the vc1902 is not virtualized.

File: lib/Dialect/AIE/Transforms/AIECanonicalizeDevice.cpp [hunk 2]
```diff
@@ -27,9 +27,9 @@ struct AIECanonicalizeDevicePass
 
     ModuleOp moduleOp = getOperation();
 
-    for (auto deviceOp : moduleOp.getOps<AIETarget>()) {
+    for (auto deviceOp : moduleOp.getOps<DeviceOp>()) {
```
Comment (stephenneuendorffer) [inline lib/Dialect/AIE/Transforms/AIECanonicalizeDevice.cpp:30]: Why did you need this?

File: include/aie/Dialect/AIE/IR/AIETargetModel.h [hunk 3]
```diff
@@ -53,8 +53,12 @@ using TileID = struct TileID {
 };
 
 class AIETargetModel {
+  bool virtualized;
+
 public:
-  AIETargetModel() = default;
+  AIETargetModel(bool virtualized = true) : virtualized(virtualized) {}
```
Comment (stephenneuendorffer) [inline include/aie/Dialect/AIE/IR/AIETargetModel.h:59]: This seems insufficient, since we actually have many potential 'virtual' targets (for instance, of different column widths.

**Review by stephenneuendorffer — CHANGES_REQUESTED** [review]

File: lib/Dialect/AIE/Transforms/AIECanonicalizeDevice.cpp (hunk 2 above)
Comment (makslevental) [reply lib/Dialect/AIE/Transforms/AIECanonicalizeDevice.cpp:30]: I removed that interface - there was only one op that implemented it `AIEDeviceOp`

**Comment by stephenneuendorffer** [comment]
Obsolete with #1443
